### PR TITLE
[Merged by Bors] - feat(AlgebraicGeometry/Morphisms): finite morphisms

### PR DESCRIPTION
**Pull request by chrisflav**
This PR defines finite morphisms of schemes.

---

- [x] depends on: #18356 

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/AlgebraicGeometry/Morphisms/Affine.lean [hunk 1]
```diff
@@ -45,6 +45,12 @@ the preimage of any affine open subset of `Y` is affine. -/
 class IsAffineHom {X Y : Scheme} (f : X ⟶ Y) : Prop where
   isAffine_preimage : ∀ U : Y.Opens, IsAffineOpen U → IsAffineOpen (f ⁻¹ᵁ U)
 
+/-- Alternative of `isAffineHom_iff` with bundled affine opens. -/
+lemma isAffineHom_iff' {X Y : Scheme.{u}} (f : X ⟶ Y) :
+    IsAffineHom f ↔ ∀ (U : Y.affineOpens), IsAffineOpen (f ⁻¹ᵁ U) := by
+  rw [isAffineHom_iff, Subtype.forall]
+  rfl
```
Comment (erdOne) [inline Mathlib/AlgebraicGeometry/Morphisms/Affine.lean:52]: Do we still need this?

Comment (chrisflav) [reply Mathlib/AlgebraicGeometry/Morphisms/Affine.lean]: No, good catch. This is an oversight.

**Comment by erdOne** [comment]
This looks good to me but there's a merge conflict.

**Comment by erdOne** [comment]
Thanks!
maintainer merge

File: Mathlib/AlgebraicGeometry/Morphisms/Finite.lean [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+/-- A morphism of schemes `X ⟶ Y` is finite if
+the preimage of any affine open subset of `Y` is affine and the induced ring
+hom is finite. -/
+@[mk_iff]
+class IsFinite {X Y : Scheme} (f : X ⟶ Y) extends IsAffineHom f : Prop where
+  finite_app (U : Y.Opens) (hU : IsAffineOpen U) : (f.app U).Finite
+
+namespace IsFinite
+
+instance : HasAffineProperty @IsFinite (affineAnd RingHom.Finite) := by
```
Comment (erdOne) [inline Mathlib/AlgebraicGeometry/Morphisms/Finite.lean:41]:
```suggestion
instance : HasAffineProperty @IsFinite fun X Y f _ ↦ IsAffine X ∧ RingHom.Finite (f.app ⊤) := by
```
Oops missed this.
I don't think we should expose `affineAnd` to end users.

**Comment by riccardobrasca** [comment]
Thanks!

bors d+

**Comment by chrisflav** [comment]
Thanks for the reviews!

bors r+

**Comment by riccardobrasca** [comment]
Can you please fix the conflict and merge it? Thanks!

**Comment by chrisflav** [comment]
Thanks for the heads up Riccardo! I didn't realize the merge did not go through.

bors r+

**Pull request closed without merging**